src/org/daisy/dotify/formatter/impl/core/TableOfContentsImpl.java

```diff
@@ -97,30 +133,118 @@ public void startEntry(String refId) {
 	
 	@Override
 	public void endEntry() {
+        if (!inTocEntry) {
+            throw new RuntimeException("Unexpected end of toc-entry");
+        }
 		inTocEntry = false;
 	}
 
+    @Override
+    public FormatterCoreImpl getEntryOnResumed() {
+        assertInTocEntryOnResumed();
+        return tocEntryOnResumedBlocks;
+    }
+
+    @Override
+	public void startEntryOnResumed(String range) {
+		if (inTocEntry || inTocEntryOnResumed) {
+			throw new RuntimeException("toc-entry and toc-entry-on-resumed may not be nested");
+		}
+		inTocEntryOnResumed = true;
+        hasTocEntryOnResumed = true;
+        
+        String startRefId = null;
+        String endRefId = null;
+        /**
+         * parse the range attribute
+         * the pattern matches only if the range is in one of these forms:
+         * [startRefId,endRefId] (unsupported) or [startRefId,endRefId) or [startRefId,)
+         * if it matches, it returns two groups: the first one is startRefId and
+         * the second one is endRefId followed by either a ']' or a ')' character
+         */
+        Pattern p = Pattern.compile("^\\[([^,\\[\\]\\)]+),([^,\\[\\]\\)]+\\]|[^,\\[\\]\\)]*\\))$");
```

Comment:
I had a hard time to read this regex as there were so many special chars. Perhaps you could simplify it to something like "^\\[(.*),(.*)\\)$" and if you don't get a match, you can check for ending of ] and then throw the exception with either incorrect range data or unsupported range data.

Reply:
This regex is the same as in the OBFL specification, see https://github.com/mtmse/obfl/blob/master/src/validation/obfl.rng line 841.
I would suggest to use the same regex here as in the OBFL specification and not to use a different regex. So let's leave it as it is. But I think it is a good idea to refer to the OBFL - I will add that to the comment.